File: js/board.js
/* jshint esversion: 6, browser: true, devel: true */
/* global BigNumber, invokeBlockCountListeners, Dialog, COLOR_POISON, mouseListeners, intToRGB, rgbToInt, toGLX, toGLY, COLOR_RED, COLOR_ORANGE, COLOR_GREEN, COLOR_BLUE, toScreenX, toScreenY, Block, VISIBLE_WIDTH, VISIBLE_HEIGHT, Data, CUBE_MESH */

let blinkTime = 0;
let blinkEnd = 0;
let blinkR = 0;
let blinkG = 0;
let blinkB = 0;
let defaultGray = 0.95;
let lightColor;

const Board = {
   // Board grid standard is 14 rows by 10 columns
   get ROWS() {
      return 14;
   },
   get COLUMNS() {
      return 10;
   },
   get SPACING() {
      return 15;
   },
   get BLOCK_WIDTH() {
      return 30;
   },
   get GRID_PADDING() {
      return this.BLOCK_WIDTH;
   },
   get FRAME_THICKNESS() {
      return this.BLOCK_WIDTH / 2;
   },

   boardCenter: {
      x: VISIBLE_WIDTH / 2,
      y: VISIBLE_HEIGHT / 2
   },

   queueFillingEnabled: true,
   fillRate: 2 * 100, // The rate at which cubes fill in units per second
   pendingFillTime: 0,
   dropQueue: function () {
      let boardFull = true;
      for (let i = 0; i < this.COLUMNS; i++)
         if (this.board[0][i] === null) {
            this.board[0][i] = this.queue[i];
            this.queue[i] = null;
            boardFull = false;
         }
      if (boardFull) this.pendingFillTime = 0;
   },
   fillQueue: function (delta) {
      if (!this.queueFillingEnabled)
         return;

      this.pendingFillTime += delta;
      let shouldDrop = true;
      for (let i = 0; i < this.COLUMNS; i++) {

         if (!this.queue[i])
            this.queue[i] = new Block();
         let block = this.queue[i];

         if ((i === 0 || this.queue[i - 1].isFull()) && !block.isFull()) {
            let f = this.queue[i].fill;
            this.queue[i].fill += this.fillRate * this.pendingFillTime;
            this.pendingFillTime -= (this.queue[i].fill - f) / this.fillRate;
            shouldDrop = false;
         }
      }
      if (shouldDrop) {
         this.dropQueue();
      }
   },
   queue: [],
   board: [],
   getGridCode: function () {
      let code = "";
      for (let r = 0; r < this.ROWS; r++)
         for (let c = 0; c < this.COLUMNS; c++) {
            if (!this.board[r][c])
               code += "0";
            else
               switch (this.board[r][c].color) {
                  case COLOR_RED:
                     code += "r";
                     break;
                  case COLOR_ORANGE:
                     code += "o";
                     break;
                  case COLOR_GREEN:
                     code += "g";
                     break;
                  case COLOR_BLUE:
                     code += "b";
                     break;
               }
         }
      return code;
   },
   dumpBlocks: [],

   init: function (ctx2d, boardCode) {
      this.ctx2d = ctx2d;
      for (let r = 0; r < this.ROWS; r++) {
         this.board[r] = [];
         for (let c = 0; c < this.COLUMNS; c++) {
            let codeIndex = r * this.COLUMNS + c;
            let color;
            if (boardCode && (color = boardCode[codeIndex]) !== '0') {
               this.board[r][c] = new Block(color);
               continue;
            }
            this.board[r][c] = null;
         }
      }

      for (let i = 0; i < this.COLUMNS; i++)
         this.queue[i] = null;

      mouseListeners.push(this);
   },

   logic: function (delta) {

      let isBlinkColor = blinkR !== 0 || blinkG !== 0 || blinkB !== 0;
      if (isBlinkColor && (blinkTime < blinkEnd || blinkEnd === 0))
         blinkTime += delta * 10;
      else blinkTime = blinkEnd;

      let foo = Math.abs(Math.sin(blinkTime)) * 3 / 4;
      let r = foo * blinkR + (1 - foo) * defaultGray;
      let g = foo * blinkG + (1 - foo) * defaultGray;
      let b = foo * blinkB + (1 - foo) * defaultGray;
      lightColor = rgbToInt(r, g, b);

      // Do board logic
      // Start from the bottom row and move up
      for (let r = this.ROWS - 1; r >= 0; r--) {
         for (let c = 0; c < this.COLUMNS; c++) {
            let block = this.board[r][c];
            if (block !== null) {
               // If a block here exists, calculate physics on
               // this block
               block.blockLogic(delta, r, c);

               // If a poison block reaches the bottom, remove it
               // (and inflict punishment >:))
               if (r === this.ROWS - 1 && block.color === COLOR_POISON) {
                  this.pushBlockToDump(r, c);
               }
            } else if (r > 0) {
               // If this slot doesn't have a block and this
               // isn't the first row, then move the block
               // above this slot to this slot
               for (let i = r; i >= 0; i--) {
                  let blockAbove = this.board[i][c];
                  if (blockAbove !== null) {
                     this.board[r][c] = blockAbove;
                     this.board[i][c] = null;
                     break;
                  }
               }
            }
         }
      }

      // Queue logic
      this.fillQueue(delta);

      for (let i = 0; i < this.COLUMNS; i++)
         if (this.queue[i])
            this.queue[i].blockLogic(delta, -1, i);


      // Do trash block logic
      this.dumpBlocks.forEach((item, index, array) => {
         item.blockLogic(delta, item.row, item.col);
         if (item.y >= item.destY && !item.falling) {
            // (For poison blocks)
            // Make the block fall if it is at or past it's
            // destination y position
            item.fall();
         }
         if (item.gone)
            this.countBlock(this.dumpBlocks.splice(index, 1)[0]);
      });

   },

   render: function (gl, programInfo) {
      this.renderBoardFrame(gl, programInfo);
      this.renderGrid(gl, programInfo);
      this.renderQueue(gl, programInfo);
      this.renderDumpBlocks(gl, programInfo);
   },
   renderQueue: function (gl, programInfo) {
      // Render the queue
      for (let i = 0; i < this.COLUMNS; i++) {
         let block = this.queue[i];
         if (block !== null)
            block.renderBlock(gl, programInfo);
      }
   },
   renderGrid: function (gl, programInfo) {
      for (let r = 0; r < this.ROWS; r++) {
         for (let c = 0; c < this.COLUMNS; c++) {
            let block = this.board[r][c];
            if (block !== null)
               block.renderBlock(gl, programInfo);
         }
      }
   },
   renderDumpBlocks: function (gl, programInfo) {
      // Render any falling blocks
      this.dumpBlocks.forEach((item, index, array) => {
         item.renderBlock(gl, programInfo);
      });
   },
   renderBoardFrame: function (gl, programInfo) {
      // Declare a shorter name for frame thickness
      let t = this.FRAME_THICKNESS;

      CUBE_MESH.setColor(0xffffff, gl, programInfo);

      // Left side of frame
      let lw = t;
      let lh = Board.height + this.GRID_PADDING;
      let lx = this.boardCenter.x - this.width / 2 - this.GRID_PADDING - t;
      let ly = this.boardCenter.y - this.height / 2;
      CUBE_MESH.render(gl, lx, ly, 0, lw, lh, t);

      // Right side of frame
      let rw = lw;
      let rh = lh;
      let rx = this.boardCenter.x + this.width / 2 + this.GRID_PADDING;
      let ry = ly;
      CUBE_MESH.render(gl, rx, ry, 0, rw, rh, t);

      //  Bottom of the frame
      let bh = t;
      let bw = Board.width + t * 2 + this.GRID_PADDING * 2;
      let bx = lx;
      let by = this.boardCenter.y + this.height / 2 + this.GRID_PADDING;
      CUBE_MESH.render(gl, bx, by, 0, bw, bh, t);

      // Lights
      CUBE_MESH.setColor(lightColor, gl, programInfo);
      CUBE_MESH.render(gl, lx, ly - t, 0, t, t, t);
      CUBE_MESH.render(gl, rx, ry - t, 0, t, t, t);
   },
   blinkLights(color, count = 2) {
      color = intToRGB(color);

      blinkR = color.r;
      blinkG = color.g;
      blinkB = color.b;

      blinkEnd = Math.PI * count;
      blinkTime = 0;
   },

   /************************************************
    * Transfers the block at the given row and
    * column to the array of falling (dump) blocks.
    ************************************************/
   pushBlockToDump: function (row, col) {
      this.dumpBlocks.push(this.board[row][col]);
      this.board[row][col] = null;
   },
   countBlock: function (block) {
      if (block.color === COLOR_POISON)
         return;

      Data.currentBlocks++;
      switch (block.color) {
         case COLOR_RED:
            Data.lifetimeBlocks.red++;
            break;
         case COLOR_ORANGE:
            Data.lifetimeBlocks.orange++;
            break;
         case COLOR_GREEN:
            Data.lifetimeBlocks.green++;
            break;
         case COLOR_BLUE:
            Data.lifetimeBlocks.blue++;
            break;
      }
   },

   toGridX(col) {
      return this.boardCenter.x - this.width / 2 + (col * (this.BLOCK_WIDTH + this.SPACING));
   },
   toGridY(row) {
      let top = this.boardCenter.y - this.height / 2;
      if (row == -1) return top;
      return top +
         // Add a row and spacing to accomodate the queue
         (row + 1) * (this.BLOCK_WIDTH + this.SPACING) +
         this.SPACING;
   },

   get width() {
      return this.COLUMNS * (this.BLOCK_WIDTH + this.SPACING) - this.SPACING;
   },
   get height() {
      // Here we include the block queue at the top, which has
      // double the spacing below it.
      return (this.ROWS + 1) * (this.BLOCK_WIDTH + this.SPACING);
   },

   /************************************************
    * REMOVE IF MATCHING NEIGHBORS
    * Kicks a block and any matching neighbors out 
    * of the board grid if said matching neighbor(s)
    * exist. Returns true if so, and false otherwise.
    ************************************************/
   removeIfMatchingNeighbors(row, col) {
      let block = this.board[row][col];
      if (!block)
         return false;

      let hadMatch = false;

      if (row > 0 && this.matchWithNeighbor(row, col, row - 1, col))
         hadMatch = true;
      if (row < this.ROWS - 1 && this.matchWithNeighbor(row, col, row + 1, col))
         hadMatch = true;

      if (col > 0 && this.matchWithNeighbor(row, col, row, col - 1))
         hadMatch = true;
      if (col < this.COLUMNS - 1 && this.matchWithNeighbor(row, col, row, col + 1))
         hadMatch = true;

      if (hadMatch) {
         this.pushBlockToDump(row, col);
         return true;
      }
      return false;
   },

   /************************************************
    * MATCH WITH NEIGHBOR
    * Matches two blocks together given their
    * rows and columns. If the blocks match, the
    * neighboring blocks will be checked for more
    * matching neighbors, and the function will 
    * return true.
    ************************************************/
   matchWithNeighbor: function (blockToRemoveRow, blockToRemoveColumn, neighborRow, neighborColumn) {
      let blockToRemove = this.board[blockToRemoveRow][blockToRemoveColumn];
      let neighbor = this.board[neighborRow][neighborColumn];

      if (!blockToRemove || !neighbor) return false;

      if (neighbor.color === blockToRemove.color) {
         blockToRemove.fall();
         // If this block has already been kicked off the board, don't bother checking
         // it for more matching neighbors
         if (!neighbor.falling)
            this.removeIfMatchingNeighbors(neighborRow, neighborColumn);
         return true;
      }
      return false;
   },
   onClick: function (mx, my) {
      let slotSize = this.BLOCK_WIDTH + this.SPACING;
      let row = Math.floor((my - this.boardCenter.y + (this.height / 2) - this.SPACING) / slotSize - 1);
      let col = Math.floor((mx - this.boardCenter.x + (this.width / 2)) / slotSize);

      if (row >= 0 && row < this.ROWS && col >= 0 && col < this.COLUMNS)
         if (this.removeIfMatchingNeighbors(row, col)) {
            Data.lifetimeClicks.successful++;
         } else {
            Data.lifetimeClicks.failed++;
         }
   }
};

File: src/board.rs
const DEFAULT_GRAY: f32 = 0.95;
const BLOCK_WIDTH: f32 = 30;
const BLOCK_GAP: i32 = 15;
const FRAME_THICKNESS: f32 = BLOCK_WIDTH / 2.0;
const ROWS: i32 = 14;
const COLUMNS: i32 = 10;

struct Board {
    blink_time: f32,
    blink_end: f32,
    blink_rgb: [u8; 3],
    light_color: f32,

    board_center: [f32; 2],
    y_offset: f32,

    fill_rate: f32,
    queue: [Option<Block>; COLUMNS],
    block_grid: [[Option<Block>; ROWS]; COLUMNS],

    pending_fill_time: f32,

    dump_blocks: Vec<Block>,
}

impl Default for Board {
    fn default() -> Self {
        Self {
            blink_time: 0.0,
            blink_end: 0.0,
            blink_rgb: [0, 0, 0],
            light_color: 0.0,

            board_center: [0, 0],

            y_offset: 0.0,

            fill_rate: 2 * 100, // The rate at which cubes fill in units per second (100 units per
            // block)
            queue: [None; COLUMNS],
            block_grid: [[None; ROWS]; COLUMNS],

            pending_fill_time: 0,

            dump_blocks: Vec::new(),
        }
    }
}

impl Board {
    /// TODO: old code.
    fn drop_queue(&self) {
        let mut board_full = true;
        for i in 0..self.COLUMNS {
            if self.board[0][i].is_none() {
                self.board[0][i] = self.queue[i];
                self.queue[i] = None;
                board_full = false;
            }
        }
        if board_full {
            self.pending_fill_time = 0
        };
    }

    /// TODO: old code.
    fn fill_queue(&self, delta: f32) {
        if !global_block_production_enabled {
            return;
        }

        self.pending_fill_time += delta;
        let should_drop = true;
        for i in 0..COLUMNS {
            if !self.queue[i] {
                self.queue[i] = Block::new();
            }
            let block = self.queue[i];

            if (i == 0 || self.queue[i - 1].is_full()) && !block.is_full() {
                let f = self.queue[i].fill;
                self.queue[i].fill += self.fill_rate * self.pending_fill_time;
                self.pending_fill_time -= (self.queue[i].fill - f) / self.fill_rate;
                should_drop = false;
            }
        }
        if should_drop {
            self.drop_queue();
        }
    }

    /// TODO: old code.
    fn get_grid_code(&self) {
        let code = "";
        for r in 0..ROWS {
            for c in 0..COLUMNS {
                if !self.board[r][c] {
                    code += "0";
                } else {
                    match self.board[r][c].color {
                        COLOR_RED => code += "r",
                        COLOR_ORANGE => code += "o",
                        COLOR_GREEN => code += "g",
                        COLOR_BLUE => code += "b",
                    }
                }
            }
        }
        code
    }

    /// TODO: old code.
    fn init<CanvasContext>(
        ctx_2d: CanvasContext,
        board_code: &str,
        mouse_listeners: &mut Listeners,
    ) -> Self {
        let grid = [[None; ROWS]; COLUMNS];
        let queue = [None; COLUMNS];
        for r in 0..ROWS {
            for c in 0..COLUMNS {
                let code_index = r * COLUMNS + c;
                let color;
                if board_code && (color = board_code[code_index]) != "0" {
                    grid[r][c] = Some(Block::new(color));
                    continue;
                }
                grid[r][c] = None;
            }
        }

        for i in 0..COLUMNS {
            queue[i] = None;
        }

        mouse_listeners.push(self);

        todo!()
    }

    /// TODO: old code.
    fn logic(delta: f32) {
        let is_blink_color = blinkR != 0 || blinkG != 0 || blinkB != 0;
        if (is_blink_color && (blinkTime < blinkEnd || blinkEnd == 0)) {
            blinkTime += delta * 10;
        } else {
            blinkTime = blinkEnd;
        }

        let foo = Math.abs(Math.sin(blinkTime)) * 3 / 4;
        let r = foo * blinkR + (1 - foo) * defaultGray;
        let g = foo * blinkG + (1 - foo) * defaultGray;
        let b = foo * blinkB + (1 - foo) * defaultGray;
        lightColor = rgbToInt(r, g, b);

        // Do board logic
        // Start from the bottom row and move up
        for r in (self.ROWS - 1)..=0 {
            for c in 0..COLUMNS {
                let block = self.board[r][c];
                if (block != null) {
                    // If a block here exists, calculate physics on
                    // self block
                    block.blockLogic(delta, r, c);

                    // If a poison block reaches the bottom, remove it
                    // (and inflict punishment >:))
                    if (r == self.ROWS - 1 && block.color == COLOR_TOXIC) {
                        self.pushBlockToDump(r, c);
                    }
                } else if (r > 0) {
                    // If this slot doesn't have a block and this
                    // isn't the first row, then move the block
                    // above this slot to this slot
                    for i in r..=0 {
                        let block_above = self.board[i][c];
                        if (block_above != null) {
                            self.board[r][c] = block_above;
                            self.board[i][c] = null;
                            break;
                        }
                    }
                }
            }
        }

        // Queue logic
        self.fillQueue(delta);

        for i in 0..COLUMNS {
            if (self.queue[i]) {
                self.queue[i].blockLogic(delta, -1, i);
            }
        }

        // Do trash block logic
        for item in self.dump_blocks {
            item.blockLogic(delta, item.row, item.col);
            if (item.y >= item.destY && !item.falling) {
                // (For poison blocks)
                // Make the block fall if it is at or past it's
                // destination y position
                item.fall();
            }
            if (item.gone) {
                self.countBlock(self.dump_blocks.splice(index, 1)[0]);
            }
        }
    }

    /// TODO: old code.
    fn render<Gl>(&self, gl: Gl, program_info: ProgramInfo, y_offset: f32) {
        self.y_offset_ = y_offset;
        self.renderBoardFrame(gl, program_info, y_offset);
        self.renderGrid(gl, program_info, y_offset);
        self.renderQueue(gl, program_info, y_offset);
        self.renderDumpBlocks(gl, program_info, y_offset);
    }

    /// TODO: old code.
    fn render_queue<Gl>(gl: Gl, program_info: ProgramInfo, y_offset: f32) {
        // Render the queue
        for i in 0..COLUMNS {
            let block = self.queue[i];
            if (block.is_some()) {
                block.renderBlock(gl, program_info, y_offset);
            }
        }
    }

    /// TODO: old code.
    fn render_grid<Gl>(gl: Gl, program_info: ProgramInfo, y_offset: f32) {
        for r in 0..ROWS {
            for c in 0..COLUMNS {
                let block = self.board[r][c];
                if (block.is_some()) {
                    block.renderBlock(gl, program_info, y_offset);
                }
            }
        }
    }

    /// TODO: old code.
    fn render_dump_blocks<Gl>(gl: Gl, program_info: ProgramInfo, y_offset: f32) {
        // Render any falling blocks
        for item in dump_blocks {
            item.renderBlock(gl, program_info, y_offset);
        }
    }

    /// TODO: old code.
    fn render_board_frame<Gl>(gl: Gl, program_info: ProgramInfo, y_offset: f32) {
        // Declare a shorter name for frame thickness
        let t = self.FRAME_THICKNESS;

        cube_mesh.set_color(0xffffff, gl, program_info);

        // Left side of frame
        let lw = t;
        let lh = Board.height + self.GRID_PADDING;
        let lx = self.boardCenter.x - self.width / 2 - self.GRID_PADDING - t;
        let ly = self.boardCenter.y - self.height / 2 + y_offset;
        cube_mesh.render(gl, lx, ly, 0, lw, lh, t);

        // Right side of frame
        let rw = lw;
        let rh = lh;
        let rx = self.boardCenter.x + self.width / 2 + self.GRID_PADDING;
        let ry = ly;
        cube_mesh.render(gl, rx, ry, 0, rw, rh, t);

        //  Bottom of the frame
        let bh = t;
        let bw = Board.width + t * 2 + self.GRID_PADDING * 2;
        let bx = lx;
        let by = self.boardCenter.y + self.height / 2 + self.GRID_PADDING + y_offset;
        cube_mesh.render(gl, bx, by, 0, bw, bh, t);

        // Lights
        cube_mesh.set_color(lightColor, gl, program_info);
        cube_mesh.render(gl, lx, ly - t, 0, t, t, t);
        cube_mesh.render(gl, rx, ry - t, 0, t, t, t);
    }

    /// TODO: old code.
    fn blink_lights(color: i32, count: u8) {
        color = intToRGB(color);

        blinkR = color.r;
        blinkG = color.g;
        blinkB = color.b;

        blinkEnd = Math.PI * count;
        blinkTime = 0;
    }

    /// Transfers the block at the given row and
    /// column to the array of falling (dump) blocks.
    ///
    /// TODO: old code.
    fn push_block_to_dump(row: i32, col: i32) {
        self.dump_blocks.push(self.board[row][col]);
        self.board[row][col] = null;
    }

    /// TODO: old code.
    fn count_block(block: &Block, data: &mut Data) {
        if (block.color == COLOR_TOXIC) {
            data.lifetime_blocks_by_color.toxic += 1;
        }

        data.current_blocks += 1;
        match (block.color) {
            COLOR_RED => data.lifetime_blocks_by_color.red += 1,
            COLOR_ORANGE => data.lifetime_blocks_by_color.orange += 1,
            COLOR_GREEN => data.lifetime_blocks_by_color.green += 1,
            COLOR_BLUE => data.lifetime_blocks_by_color.blue += 1,
        }
    }

    /// TODO: old code.
    fn to_grid_x(col: i32) {
        self.boardCenter.x - self.width / 2 + (col * (self.BLOCK_WIDTH + self.SPACING))
    }

    /// TODO: old code.
    fn to_grid_y(row: i32) {
        let top = self.boardCenter.y - self.height / 2;
        if (row == -1) {
            top
        } else {
            top +
                // Add a row and spacing to accomodate the queue
                (row + 1) * (self.BLOCK_WIDTH + self.SPACING) +
                self.SPACING;
        }
    }

    /// TODO: old code.
    fn get_width() {
        self.COLUMNS * (self.BLOCK_WIDTH + self.SPACING) - self.SPACING
    }

    /// TODO: old code.
    fn get_height() {
        // Here we include the block queue at the top, which has
        // double the spacing below it.
        (self.ROWS + 1) * (self.BLOCK_WIDTH + self.SPACING)
    }

    /// Kicks a block and any matching neighbors out
    /// of the board grid if said matching neighbor(s)
    /// exist. s true if so, and false otherwise.
    ///
    /// TODO: old code.
    fn remove_if_matching_neighbors(&self, row: i32, col: i32) -> bool {
        let block = self.board[row][col];

        if (!block) {
            return false;
        }

        let had_match = false;

        if (row > 0 && self.match_with_neighbor(row, col, row - 1, col)) {
            had_match = true;
        }
        if (row < self.ROWS - 1 && self.match_with_neighbor(row, col, row + 1, col)) {
            had_match = true;
        }

        if (col > 0 && self.match_with_neighbor(row, col, row, col - 1)) {
            had_match = true;
        }
        if (col < self.COLUMNS - 1 && self.match_with_neighbor(row, col, row, col + 1)) {
            had_match = true;
        }

        if (had_match) {
            self.push_block_to_dump(row, col);
            true
        } else {
            false
        }
    }

    /// Matches two blocks together given their rows and columns. If the blocks match, the
    /// neighboring blocks will be checked for more matching neighbors, and the fn will
    /// true.
    ///
    /// TODO: old code.
    fn match_with_neighbor(
        &self,
        block_to_remove_row: i32,
        block_to_remove_column: i32,
        neighbor_row: i32,
        neighbor_column: i32,
    ) -> bool {
        let block_to_remove = self.board[block_to_remove_row][block_to_remove_column];
        let neighbor = self.board[neighbor_row][neighbor_column];

        if (!block_to_remove || !neighbor) {
            return false;
        }

        if (neighbor.color == block_to_remove.color) {
            block_to_remove.fall();
            // If self block has already been kicked off the board, don't bother checking
            // it for more matching neighbors
            if (!neighbor.falling) {
                self.remove_if_matching_neighbors(neighbor_row, neighbor_column);
            }
            true
        } else {
            false
        }
    }

    /// TODO: old code.
    fn on_click(&self, mx: i32, my: i32, data: &mut Data) {
        let block_cell_size = BLOCK_WIDTH + BLOCK_GAP;
        let row = ((my - self.board_center.y - self.y_offset_ + (self.height / 2)
            - self.BLOCK_GAP)
            / block_cell_size
            - 1)
        .floor();
        let col = ((mx - self.board_center.x + (self.width / 2)) / block_cell_size).floor();

        if (row >= 0 && row < self.ROWS && col >= 0 && col < self.COLUMNS) {
            if (self.remove_if_matching_neighbors(row, col)) {
                data.lifetime_clicks.successful += 1;
            } else {
                data.lifetime_clicks.failed += 1;
            }
        }
    }
}
